### PR TITLE
#11032 Support jira personal access token

### LINKED ISSUE #11032: [Jira Commit Hint Plugin] Support jira personal access token.
### Feature description

With the policy changed, our company deined jira basic auth. So i create one `Personal access token` on Jira, but it didn't work with current JiraCommitHintPlugin.
So can we make some improment on JiraCommitHintPlugin?

My make some code changed like below:
image
image

image
image

If use JWT, then take `API token/Password`.

### Environment

- Git Extensions 4.1.0.16698
- Build 1fe52a137acef1fbd2b653dea263b44ad5844177
- Git 2.41.0.windows.1
- Microsoft Windows NT 10.0.22621.0
- .NET 6.0.16
- DPI 96dpi (no scaling)
- Portable: False
- Microsoft.WindowsDesktop.App Versions

```
    Microsoft.WindowsDesktop.App 6.0.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 6.0.16 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 7.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

### PR DESCRIPTION
**Pull request by beiifeng**
Fixes #11032

## Proposed changes

- JiraCommitHintPlugin add `Auth type` to support Personal access token.

### Before

image

### After

image
image

## Test methodology 

image
image 

## Test environment(s) 

- Git 2.41.0.windows.1
- Microsoft Windows NT 10.0.22621.0

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs [hunk 1]
```diff
@@ -25,11 +27,14 @@ public class JiraCommitHintPlugin : GitPluginBase, IGitPluginForRepository
         private static readonly TranslationString EmptyQueryResultCaption = new("First Task Preview");
 
         private const string DefaultFormat = "{Key} {Summary}";
+        private const string BASIC = "Basic";
+        private const string JWT = "JWT";
```
Comment (gerhardol) [inline Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs:31]: Present something more user friendly to the user, like Personal Access Token.

File: Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs [hunk 2]
```diff
@@ -25,11 +27,14 @@ public class JiraCommitHintPlugin : GitPluginBase, IGitPluginForRepository
...
         private Jira? _jira;
         private string? _query;
         private string _stringTemplate = DefaultFormat;
         private readonly BoolSetting _enabledSettings = new("Jira hint plugin enabled", false);
         private readonly StringSetting _urlSettings = new("Jira URL", @"https://jira.atlassian.com");
+        private readonly ChoiceSetting _authTypeSettings = new("Auth type", new List<string> { BASIC, JWT });
```
Comment (gerhardol) [inline Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs:37]: Hide username if PAT is used, should change label for pat/pw too.
Suggest to add PAT as the default, unless username is added already

**Review by gerhardol — COMMENTED** [review]
Have not run

File: Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs (hunk 2 above)
Comment (beiifeng) [reply Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs:37]: Yeah, I'd like to make some optimize, before next tuesday.

File: Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs (hunk 1 above)
Comment (RussKie) [inline Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs:31]: Please use sentence case for consts too.

```suggestion
        private const string AuthTypeBasic = "Basic";
        private const string AuthTypeJwt = "JWT";
```

File: Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs [hunk 3]
```diff
@@ -282,6 +290,17 @@ private async Task<JiraTaskDTO[]> GetMessageToCommitAsync(Jira jira, string quer
             }
         }
 
+        private Jira createJiraClient(string url, string username, string password, string authType)
```
Comment (RussKie) [inline Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs:293]: [styling] Methods names must be in PascalCase (unless those are Windows Forms controls' event handlers).

File: Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs [hunk 4]
```diff
@@ -293,5 +312,12 @@ public JiraTaskDTO(string title, string text)
                 Text = text;
             }
         }
+
+        private class JiraJWTRestClient : JiraRestClient
```
Comment (RussKie) [inline Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs:316]:
```suggestion
        private class JiraJwtRestClient : JiraRestClient
```

File: Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs [hunk 5]
```diff
@@ -282,6 +290,17 @@ private async Task<JiraTaskDTO[]> GetMessageToCommitAsync(Jira jira, string quer
             }
         }
 
+        private Jira createJiraClient(string url, string username, string password, string authType)
+        {
+            if (JWT.Equals(authType))
```
Comment (RussKie) [inline Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs:295]: This is more idiomatic:
```suggestion
            if (JWT == authType)
```

**Comment by gerhardol** [comment]
Works for me.
If it is complicated to hide the username etc, it should be an option to only use PAT. Jira plugin was disabled in 4.0 and reactivated in 4.1, so it is not so long history for users that are affected.
Similar changes should be done for the BitBucket Server plugin (not in this PR). (I have tested the changes in a branch and plan to align the username handling with this PR, unless someone else is more interested in making those changes - I hardly use the plugin myself).

**Comment by RussKie** [comment]
Instead of hiding, make read-only?

**Comment by beiifeng** [comment]
> Instead of hiding, make read-only?
> 

For user experience,  it's better to hide it .

File: Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs [hunk 6]
```diff
@@ -27,14 +27,17 @@ public class JiraCommitHintPlugin : GitPluginBase, IGitPluginForRepository
...
+
+        // It not a good idea use this user name for Personal access token, but credentials required.
+        private const string AuthTypeJwtUserName = "#Jira.Commit.Hint:access.token$";
         private Jira? _jira;
         private string? _query;
         private string _stringTemplate = DefaultFormat;
         private readonly BoolSetting _enabledSettings = new("Jira hint plugin enabled", false);
         private readonly StringSetting _urlSettings = new("Jira URL", @"https://jira.atlassian.com");
-        private readonly ChoiceSetting _authTypeSettings = new("Auth type", new List<string> { BASIC, JWT });
+        private readonly ChoiceSetting _authTypeSettings = new("Auth type", new List<string> { AuthTypeBasic, AuthTypeJwt });
```
Comment (gerhardol) [inline Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs:40]: PAT first?

File: Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs [hunk 7]
```diff
@@ -81,6 +89,11 @@ public override IEnumerable<ISetting> GetSettings()
             _urlSettings.CustomControl = new TextBox();
             yield return _urlSettings;
 
+            _authTypeSettings.CustomControl = new ComboBox { DropDownStyle = ComboBoxStyle.DropDownList };
+            _authTypeSettings.CustomControl.Items.AddRange(_authTypeSettings.Values.ToArray());
```
Comment (RussKie) [inline Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs:93]: [nit] This creates unnecessary allocations. Suggest unrolling this into `foreach (... value in _authTypeSettings.Values) { _authTypeSettings.CustomControl.Items.Add(item); }`

File: Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs [hunk 8]
```diff
@@ -169,6 +182,43 @@ private void btnPreviewClick(object sender, EventArgs eventArgs)
             }
         }
 
+        private void AuthTypeSelectedChange(object sender, EventArgs e)
+        {
+            if ((sender is ComboBox comboBox) && _credentialsSettings.CustomControl != null)
```
Comment (RussKie) [inline Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs:187]:
```suggestion
            if ((sender is ComboBox comboBox) && _credentialsSettings.CustomControl is not null)
```

Comment (RussKie) [inline Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs:187]: Please invert with early exit to reduce nesting

File: Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs [hunk 9]
```diff
@@ -169,6 +182,43 @@ private void btnPreviewClick(object sender, EventArgs eventArgs)
...
+            {
+                var credentials = _credentialsSettings.GetValueOrDefault(Settings);
+                var authType = _authTypeSettings.ValueOrDefault(Settings);
+
+                if (comboBox.SelectedItem.ToString() == AuthTypeJwt)
+                {
+                    _credentialsSettings.CustomControl.ChangeUIMode(false, "Personal access token");
```
Comment (RussKie) [inline Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs:194]:
```suggestion
                    _credentialsSettings.CustomControl.ChangeUIMode(showUserName: false, "Personal access token");
```

File: Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs [hunk 10]
```diff
@@ -205,6 +256,12 @@ private void UpdateJiraSettings()
 
             _btnPreview.Click -= btnPreviewClick;
             _btnPreview = null;
+
+            if (_authTypeSettings.CustomControl != null)
```
Comment (RussKie) [inline Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs:260]:
```suggestion
            if (_authTypeSettings.CustomControl is not null)
```

File: Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs (hunk 6 above)
Comment (beiifeng) [reply Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs:40]: OK

File: Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs [hunk 11]
```diff
@@ -86,6 +89,9 @@ public override IEnumerable<ISetting> GetSettings()
             _urlSettings.CustomControl = new TextBox();
             yield return _urlSettings;
 
+            _authTypeSettings.CustomControl = new ComboBox { DropDownStyle = ComboBoxStyle.DropDownList };
+            _authTypeSettings.CustomControl.Items.AddRange(_authTypeSettings.Values.ToArray());
+            _authTypeSettings.CustomControl.SelectedIndexChanged += AuthTypeSelectedChange;
```
[the remark this answers is not in the document]
Comment (beiifeng) [reply Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs:94]: got it

File: Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs (hunk 7 above)
Comment (beiifeng) [reply Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs:93]: got it

File: Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs (hunk 6 above)
Comment (beiifeng) [reply Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs]: I tried to PAT first, but got some accident with migrate. If user use username/password already, we think they sholudn't make any change to use it still. But if PAT first, they have to change their settings.

**Review by RussKie — APPROVED** [review]
👍 (haven't run though)

**Comment by RussKie** [comment]
@beiifeng in the future consider not working off branched you didn't create as those branches can and do move outside your control. Consider creating your own feature branches instead.
image

**Comment by beiifeng** [comment]
> @beiifeng in the future consider not working off branched you didn't create as those branches can and do move outside your control. Consider creating your own feature branches instead.
> image
> 

Got it.Thanks all.

**Comment by RussKie** [comment]
@pmiossec @gerhardol if there are no objections I'll merge it in few days.

**Comment by gerhardol** [comment]
> if there are no objections I'll merge it in few days.

OK for me. I had a few comments about defaults (but that migration more difficult).
Had missed this patch in the environment I can test today, earliest on Thursday - but that is not necessary. (plan to use the same approach for BB plugin)

**Comment by RussKie** [comment]
Thank you

**Comment by beiifeng** [comment]
Thanks all.

**Pull request merged**